codex-rs/mcp-server/src/conversation_loop.rs

```diff
@@ -1,38 +1,147 @@
 use codex_core::protocol::ExecApprovalRequestEvent;
+use codex_core::protocol::FileChange;
 use mcp_types::RequestId;
+use tokio::sync::Mutex;
+// no streaming watch channel; streaming is toggled via set_streaming on the struct
 use tracing::error;
+use uuid::Uuid;
+
+use crate::exec_approval::handle_exec_approval_request;
+use crate::mcp_protocol::CodexEventNotificationParams;
+use crate::mcp_protocol::ConversationId;
+use crate::mcp_protocol::InitialStateNotificationParams;
+use crate::mcp_protocol::InitialStatePayload;
+use crate::mcp_protocol::NotificationMeta;
+use crate::outgoing_message::OutgoingMessageSender;
+use crate::patch_approval::handle_patch_approval_request;
 
-pub async fn run_conversation_loop(
+/// Conversation struct that owns the Codex session and all per-conversation state.
+pub(crate) struct Conversation {
     codex: Arc<Codex>,
+    session_id: Uuid,
     outgoing: Arc<OutgoingMessageSender>,
     request_id: RequestId,
-) {
-    let request_id_str = match &request_id {
-        RequestId::String(s) => s.clone(),
-        RequestId::Integer(n) => n.to_string(),
-    };
-
-    // Stream events until the task needs to pause for user interaction or
-    // completes.
-    loop {
-        match codex.next_event().await {
+    state: Mutex<ConversationState>,
+}
+
+struct ConversationState {
+    running: bool,
+    streaming_enabled: bool,
+    buffered_events: Vec<CodexEventNotificationParams>,
+    pending_elicitations: Vec<PendingElicitation>,
+}
+
+impl Conversation {
+    pub(crate) fn new(
+        codex: Arc<Codex>,
+        outgoing: Arc<OutgoingMessageSender>,
+        request_id: RequestId,
+        session_id: Uuid,
+    ) -> Arc<Self> {
+        let conv = Arc::new(Self {
+            codex,
+            session_id,
+            outgoing,
+            request_id,
+            state: Mutex::new(ConversationState {
+                running: false,
+                streaming_enabled: false,
+                buffered_events: Vec::new(),
+                pending_elicitations: Vec::new(),
+            }),
+        });
+        // Detach a background loop tied to this Conversation
+        Conversation::spawn_loop(conv.clone());
+        conv
+    }
+
+    pub(crate) async fn set_streaming(&self, enabled: bool) {
+        if enabled {
+            let (events_snapshot, pending_snapshot) = {
+                let mut st = self.state.lock().await;
+                st.streaming_enabled = true;
+                (
+                    st.buffered_events.clone(),
+                    std::mem::take(&mut st.pending_elicitations),
+                )
+            };
+            self.emit_initial_state_with(events_snapshot).await;
+            self.drain_pending_elicitations_from(pending_snapshot).await;
+        } else {
+            let mut st = self.state.lock().await;
+            st.streaming_enabled = false;
+        }
+    }
+
+    fn spawn_loop(this: Arc<Self>) {
+        tokio::spawn(async move {
+            loop {
```

Comment:
Clone once outside the loop and then do the `while let` style thing to iterate the events. Also, I would only handle the Ok case in the helper function you're calling. Deal with the Err in the loop?